### PR TITLE
Azure Gaming Resources

### PR DESCRIPTION
**Pull request by dciborow**
## Description

If you haven't already, read the full [contribution guide](../CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

## Adding a new module

- [ ] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Updating an existing module

- [ ] This is a bug fix:
  - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
  - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
- [ ] I have run `brm validate` locally to verify the module files.
- [ ] I have run deployment tests locally to ensure the module is deployable.
- [ ] I have read the [Updating an existing module](../CONTRIBUTING.md#updating-an-existing-module) section in the contributing guide and updated the `version.json` file properly:
  - [ ] The PR contains backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`.
  - [ ] The PR contains backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] The PR contains breaking changes, and I have bumped the MAJOR version in `version.json`.

### CODE REVIEW

File: modules/azure-gaming/connected-resources/metadata.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "$schema": "https://aka.ms/bicep-registry-module-metadata-file-schema-v2#",
+  "name": "",
+  "summary": "",
+  "owner": ""
```
Comment (dciborow) [inline modules/azure-gaming/connected-resources/metadata.json:5]:
```suggestion
  "name": "Azure Gaming: Connected Resources",
  "summary": "This module is used by other Azure Gaming modules and applications to create core Azure services",
  "owner": "dciborow"
```

File: .github/workflows/push-auto-generate.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        run: |
+          dotnet tool install --global Azure.Bicep.RegistryModuleTool
+
+          cd $SAMPLEFOLDER_PATH
+          brm generate
+
+      - uses: EndBug/add-and-commit@v9
+        if: steps.tag_version.outputs.release_type
+        with:
+          message: Update Version to ${{ steps.tag_version.outputs.new_version }}
```
Comment (dciborow) [inline .github/workflows/push-auto-generate.yml:79]:
```suggestion
          message: Autogenerate Bicep Files
```

File: .github/workflows/push-auto-generate.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+          committer_name: GitHub Actions
+          committer_email: actions@github.com
+          add: modules
```
Comment (dciborow) [inline .github/workflows/push-auto-generate.yml:82]:
```suggestion
          add: modules README.md
```

File: modules/azure-gaming/connected-resources/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# Azure Gaming: Connected Resources
+
+This module is used by other Azure Gaming modules and applications to create core Azure services
+
+## Description
+
+{{ Add detailed description for the module. }}
```
Comment (dciborow) [inline modules/azure-gaming/connected-resources/README.md:7]:
```suggestion
This module is used to deploy core Azure services in Azure Gaming deployments. These resources can be configured to use availability zones in available regions.
```

File: modules/azure-gaming/connected-resources/modules/documentDB/databaseAccounts.bicep [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  properties: {
+    consistencyPolicy: consistencyPolicy[defaultConsistencyLevel]
+    locations: locations
+    databaseAccountOfferType: 'Standard'
+    enableAutomaticFailover: systemManagedFailover
+    enableMultipleWriteLocations: enableMultipleWriteLocations
+    capabilities: union([ { name: 'EnableCassandra' } ], EnableServerless ? [ { name: 'EnableServerless' } ] : [])
+  }
+}
+resource account 'Microsoft.DocumentDB/databaseAccounts@2022-05-15' existing = { name: toLower(name) }
```
Comment (dciborow) [inline modules/azure-gaming/connected-resources/modules/documentDB/databaseAccounts.bicep:92]:
```suggestion

resource account 'Microsoft.DocumentDB/databaseAccounts@2022-05-15' existing = { name: toLower(name) }
```

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -4,6 +4,9 @@
 
 If you haven't already, read the full [contribution guide](../CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.
 
+Enable GitHub Worksflows in your fork to enable auto-generation of assets with our [GitHub Action](/.github/workflows/push-auto-generate.yml).
+To trigger GitHub Actions after auto-generation, [add a GitHub PAT](https://stackoverflow.com/a/69063453/1341806) as a secret in your forked repository called `PAT`.
+
```
Comment (dciborow) [inline .github/pull_request_template.md:9]:
```suggestion
```

File: modules/azure-gaming/connected-resources/modules/resources.bicep [hunk 7]
```diff
@@ -0,0 +1,157 @@
+@description('Deployment Location')
+param location string
+
+@description('Toggle to enable or disable zone redudance.')
+param isZoneRedundant bool = true
+
+@allowed([ 'new', 'existing', 'none' ])
+param newOrExistingStorageAccount string = 'none'
+param storageAccountName string = 'store${uniqueString(resourceGroup().id, subscription().id)}'
```
Comment (Gordonby) [inline modules/azure-gaming/connected-resources/modules/resources.bicep:9]: The default value for storageAccountName is different in 3 places throughout the modules.

The CAF naming guidance suggests using st as the prefix. `st<project, app or service><###>`

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep [hunk 8]
```diff
@@ -0,0 +1,104 @@
+@description('Deployment Location')
+param location string
+param name string = 'k8-cluster'
+param agentPoolCount int = 3
+param agentPoolName string = 'k8agent'
+param vmSize string = 'Standard_L16s_v2'
+param nodeLabels string = 'horde-storage'
+param assignRole bool = false
+param dnsPrefix string = 'k8-${take(uniqueString(name), 5)}'
+param kubernetesVersion string = '1.22.11'
```
Comment (Gordonby) [inline modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:10]: The portal default for k8s version is `1.23.12`
1.22.11 is currently the oldest deployable version
image

File: modules/azure-gaming/connected-resources/modules/resources.bicep [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+  name: 'create_cluster-${uniqueString(location, resourceGroup().id, deployment().name)}'
+  params: {
+    name: contains(kubernetesParams, 'name') ? kubernetesParams.name : 'aks-${uniqueString(resourceGroup().id)}'
+    location: location
+    agentPoolCount: contains(kubernetesParams, 'agentPoolCount') ? kubernetesParams.agentPoolCount : 3
+    agentPoolName: contains(kubernetesParams, 'agentPoolName') ? kubernetesParams.agentPoolName : 'agentpool'
+    vmSize: contains(kubernetesParams, 'vmSize') ? kubernetesParams.vmSize : 'Standard_D2_v2'
+    assignRole: assignRole
+    newOrExisting: newOrExisting[newOrExistingKubernetes]
+    isZoneRedundant: isZoneRedundant && !contains(noAvailabilityZones, location)
```
Comment (Gordonby) [inline modules/azure-gaming/connected-resources/modules/resources.bicep:89]: As you've written this logic in 3 places, i'd promote it a variable to then reference.

```bicep
var locationHasAzSupport = !contains(noAvailabilityZones, location)
```

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep [hunk 10]
```diff
@@ -0,0 +1,104 @@
+@description('Deployment Location')
+param location string
+param name string = 'k8-cluster'
+param agentPoolCount int = 3
+param agentPoolName string = 'k8agent'
+param vmSize string = 'Standard_L16s_v2'
+param nodeLabels string = 'horde-storage'
```
Comment (Gordonby) [inline modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:7]: Is this default label workload specific?

**Review by Gordonby — CHANGES_REQUESTED** [review]
Taken a first pass through the modules.

**Comment by Gordonby** [comment]
I do see some similarity here with what I previously proposed in #61 (and was rejected).

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep (hunk 8 above)
Comment (alex-frankel) [reply modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:10]: Should this param not have a default value? it seems like the supported k8s version gets updated very frequently. Maybe the param description can have a link for the latest supported versions?

**Comment by alex-frankel** [comment]
> I do see some similarity here with what I previously proposed in #61 (and was rejected).

@Gordonby - I think back in April we were hoping we would be able to onboard CARML modules more easily. We are more open now to onboarding "base resource" modules. From #61, it looks like that modules was explicitly intended as a sample, which is what quickstarts if for. If it's not intended to be used as a general purpose utility, then it should live here.

**Comment by alex-frankel** [comment]
@dciborow -- just to make sure I understand, the `modules` directory in this module is only intended to be used with this module, but eventually they may "graduate" to top level modules anyone can use?

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep (hunk 8 above)
Comment (Gordonby) [reply modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:10]: > > I do see some similarity here with what I previously proposed in #61 (and was rejected).
> 
> @Gordonby - I think back in April we were hoping we would be able to onboard CARML modules more easily. We are more open now to onboarding "base resource" modules. From #61, it looks like that modules was explicitly intended as a sample, which is what quickstarts if for. If it's not intended to be used as a general purpose utility, then it should live here.

I think back then i was concerned it didn't align to a specific azure service so was trying to work out what the right directory to put it in would be.

**Comment by Gordonby** [comment]
> > I do see some similarity here with what I previously proposed in #61 (and was rejected).
> 
> @Gordonby - I think back in April we were hoping we would be able to onboard CARML modules more easily. We are more open now to onboarding "base resource" modules. From #61, it looks like that modules was explicitly intended as a sample, which is what quickstarts if for. If it's not intended to be used as a general purpose utility, then it should live here.

I think i was struggling to see what directory it would fit it since it's deploying so many services. `/samples` was a guess.

**Comment by alex-frankel** [comment]
> > > I do see some similarity here with what I previously proposed in #61 (and was rejected).
> > 
> > 
> > @Gordonby - I think back in April we were hoping we would be able to onboard CARML modules more easily. We are more open now to onboarding "base resource" modules. From #61, it looks like that modules was explicitly intended as a sample, which is what quickstarts if for. If it's not intended to be used as a general purpose utility, then it should live here.
> 
> I think i was struggling to see what directory it would fit it since it's deploying so many services. `/samples` was a guess.

Fair enough -- I think it would be appropriate in the `/app` directory if you wanted to re-open it.

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep (hunk 8 above)
Comment (dciborow) [reply modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:10]: and.... '1.22.11' disappeared from East US, so it's time to update again!

**Comment by dciborow** [comment]
> @dciborow -- just to make sure I understand, the `modules` directory in this module is only intended to be used with this module, but eventually they may "graduate" to top level modules anyone can use?

Yes.
I specifically need each of these files in our Bicep Registry so that I can avoid putting all this source code into my shared repo with Epic. Instead, using this repo, I can manage all this Azure complexity outside of that project.

But it also lets me put all my cards on the table so that @Gordonby can see what I am cooking and provide feedback as well.

Then, in the future we can also refactor/dedup based on what would be useful outside of Azure Gaming.

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep (hunk 10 above)
Comment (dciborow) [reply modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:7]:
```suggestion
param nodeLabels string = 'aks-nodes'
```

Comment (dciborow) [reply modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:7]: Yes, changing the default, but will verify a default even makes sense...

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep (hunk 8 above)
Comment (dciborow) [reply modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:10]:
```suggestion
param kubernetesVersion string = '1.23.8'
```

File: modules/azure-gaming/connected-resources/modules/storage/storageAccounts.bicep [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+resource storageAccount 'Microsoft.Storage/storageAccounts@2021-09-01' existing = {
+  scope: resourceGroup(resourceGroupName)
+  name: name
+}
+
+var keys = newOrExisting == 'new' ? listKeys(newStorageAccount.id, newStorageAccount.apiVersion) : listKeys(storageAccount.id, storageAccount.apiVersion)
+var blobStorageConnectionString = 'DefaultEndpointsProtocol=https;AccountName=${name};EndpointSuffix=${environment().suffixes.storage};AccountKey=${keys.keys[0].value}'
+
+output id string = newOrExisting == 'new' ? newStorageAccount.id : storageAccount.id
+output blobStorageConnectionString string = blobStorageConnectionString
```
[the remark this answers is not in the document]
Comment (dciborow) [reply modules/azure-gaming/connected-resources/modules/storage/storageAccounts.bicep:66]:
```suggestion
```

File: modules/azure-gaming/connected-resources/modules/keyvault/vaults.bicep [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+resource existingKeyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = if (newOrExisting == 'existing') {
+  name: name
+}
+
+resource identityRoleAssignDeployment 'Microsoft.Authorization/roleAssignments@2022-04-01' = [for rbacPolicy in rbacPolicies: if (assignRole) {
+  name: guid(rbacSecretsReaderRole, rbacPolicy.objectId)
+  scope: keyVault
+  properties: {
+    roleDefinitionId: resourceId('Microsoft.Authorization/roleDefinitions', rbacSecretsReaderRole)
```
[the remark this answers is not in the document]
Comment (dciborow) [reply modules/azure-gaming/connected-resources/modules/keyvault/vaults.bicep:50]:
```suggestion
    roleDefinitionId: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', rbacSecretsReaderRole)
```

File: modules/azure-gaming/connected-resources/modules/resources.bicep (hunk 7 above)
Comment (dciborow) [reply modules/azure-gaming/connected-resources/modules/resources.bicep:9]:
```suggestion
param storageAccountName string = 'st${uniqueString(resourceGroup().id, subscription().id)}'
```

File: modules/azure-gaming/connected-resources/modules/storage/storageAccounts.bicep [hunk 13]
```diff
@@ -0,0 +1,65 @@
+@description('Deployment Location')
+param location string
+
+param name string = uniqueString(resourceGroup().id)
```
Comment (dciborow) [inline modules/azure-gaming/connected-resources/modules/storage/storageAccounts.bicep:4]:
```suggestion
param name string = 'st${uniqueString(resourceGroup().id, subscription().id)}'
```

File: modules/azure-gaming/connected-resources/main.bicep [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+@description('AKS VM Size')
+param vmSize string = 'Standard_L16s_v2'
+
+@allowed([ 'new', 'existing', 'none' ])
+@description('Set to create a new or use an existing Storage Account')
+param newOrExistingStorageAccount string = 'none'
+
+@description('Storage Account Resource Name')
+param storageAccountName string = 'data${uniqueString(resourceGroup().id, subscription().subscriptionId)}'
```
Comment (dciborow) [inline modules/azure-gaming/connected-resources/main.bicep:28]:
```suggestion
param storageAccountName string = 'st${uniqueString(resourceGroup().id, subscription().id)}'
```

File: modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep [hunk 15]
```diff
@@ -0,0 +1,104 @@
+@description('Deployment Location')
+param location string
+param name string = 'k8-cluster'
```
Comment (dciborow) [inline modules/azure-gaming/connected-resources/modules/ContainerService/managedClusters.bicep:3]:
```suggestion
param name string = 'aks-${uniqueString(resourceGroup().id, subscription().id)}'
```

**Pull request closed without merging**